### PR TITLE
Introduce database schema versions

### PR DESCRIPTION
**Pull request by invisiblek**
This essentially does nothing for now other than creating the table we'll use in the future.

### CODE REVIEW

File: pogom/models.py [hunk 1]
```diff
@@ -457,3 +469,27 @@ def drop_tables(db):
...
+        db_ver = Versions.get(Versions.key=='schema_version').val
+
+        if db_ver < db_schema_version:
+            # This is where we'll handle migrations in the future
+            log.info("Detected database version %i, updating to %i", db_ver, db_schema_version)
+            Versions.update(val=db_schema_version).where(Versions.key=='schema_version').execute()
+        elif db_ver > db_schema_version:
+            log.error("Your database version (%i) appears to be newer than the code supports (%i).",
+                      db_ver, db_schema_version)
+            log.error("Please downgrade your code base or drop all tables in your database.")
```
Comment (Chlodochar) [inline pogom/models.py]: shouldn't this be "*upgrade* your codebase"?

**Comment by Chlodochar** [comment]
+10000. Haven't been able to test this yet, but this should've been in there since the very beginning. Apart from the comment, for looks good :+1:.

File: pogom/models.py (hunk 1 above)
Comment (invisiblek) [inline pogom/models.py]: yea, it was getting late, i'll fixit

**Comment by invisiblek** [comment]
Ok, I think i'm finally happy with this.

File: pogom/models.py [hunk 2]
```diff
@@ -449,11 +460,45 @@ def bulk_upsert(cls, data):
...
     db.connect()
-    db.drop_tables([Pokemon, Pokestop, Gym, ScannedLocation], safe=True)
+    db.drop_tables([Pokemon, Pokestop, Gym, ScannedLocation, Versions], safe=True)
     db.close()
+
+
+def verify_database_schema(db):
+    if not Versions.table_exists():
+        db.create_tables([Versions])
+        InsertQuery(Versions, {Versions.key: 'schema_version', Versions.val: 0}).execute()
```
Comment (invisiblek) [inline pogom/models.py]: hmm, the logic is off here i think

**Comment by GenuineSounds** [comment]
In the future will we be distributing updates to the database in the form of incremental sql patches instead of hardcoded python scripts?

**Comment by invisiblek** [comment]
@GenuineSounds I never thought about that. Long-term that might be a good idea.

**Comment by GenuineSounds** [comment]
@invisiblek All of the project I've used long terms always used version to version sql patches and it worked quite well. Of which I'm speaking about https://github.com/mangos/MaNGOS databases and the open source World of Warcraft databases since the beginning basically.

**Comment by invisiblek** [comment]
We'd have to take things into consideration like the fact we're supporting sqlite and mysql databases. Do we need to be concerned with syntax between the two when we're writing straight SQL? Going through peewee abstracts all that for us.

**Comment by GenuineSounds** [comment]
@invisiblek You're right about that... That's an interesting problem. We are going to need an abstraction layer if we want a common patch format. DAMN YOU SQL!!!

**Comment by invisiblek** [comment]
Android does it like this, so it can't be ALL bad =P (Android uses sqlite fwiw)

**Comment by GenuineSounds** [comment]
I'm not well versed in python, I assume the OOP concepts are the same, so I'll speak in those terms.

We have an interface for the patch classes we use and load them dynamically from a folder... by name I suppose. something like
```
interface { string from, to; status patch(db); }
```
Easy to manage updates, either load them by name or load them all and walk the update chain.

**Comment by Chlodochar** [comment]
@GenuineSounds It's a good approach, but you'll still need to automatically choose the right migration. I think for the time being a simple integer for the version with linear migrations is quite sufficient already.

**Comment by invisiblek** [comment]
let's see how this goes...

**Pull request merged**